source/index.ts

```diff
@@ -0,0 +1,315 @@
+import EventEmitter from 'eventemitter3';
+import {Queue} from './queue';
+import PriorityQueue from './priority-queue';
+import {QueueAddOptions, DefaultAddOptions, Options} from './options';
+
+type ResolveFunction<T = void> = (value?: T | PromiseLike<T>) => void;
+
+type Task<TaskResultType> =
+		| (() => PromiseLike<TaskResultType>)
+		| (() => TaskResultType);
+
+const empty = () => {};
+
+/**
+Promise queue with concurrency control.
+*/
+export default class PQueue<QueueType extends Queue<EnqueueOptionsType> = PriorityQueue, EnqueueOptionsType extends QueueAddOptions = DefaultAddOptions> extends EventEmitter<'active'> {
+	private readonly _carryoverConcurrencyCount: boolean;
+
+	private readonly _isIntervalIgnored: boolean;
+
+	private _intervalCount = 0;
+
+	private readonly _intervalCap: number;
+
+	private readonly _interval: number;
+
+	private _intervalEnd = 0;
+
+	private _intervalId?: NodeJS.Timeout;
+
+	private _timeoutId?: NodeJS.Timeout;
+
+	private _queue: QueueType;
+
+	private readonly _queueClass: new () => QueueType;
+
+	private _pendingCount = 0;
+
+	private readonly _concurrency: number;
+
+	private _paused: boolean;
+
+	private _resolveEmpty: ResolveFunction = empty;
+
+	private _resolveIdle: ResolveFunction = empty;
+
+	constructor(opt?: Options<QueueType, EnqueueOptionsType>) {
```

Comment:
`opt` => `options`

Reply:
that requires a typecast (or custom typeguard when using a method) to tell ts it can't be `undefined` anymore thus helping out since it allows us to drop the `@ts-ignore` 🎉